### PR TITLE
Blog post: Arbitrary CPU and RAM supported in ComputeService

### PR DESCRIPTION
**Pull request by ivanlomba**
Summary of new feature, the post includes:
- Description
- Providers supported
- How to create custom hardwares
- Further development

### CODE REVIEW

**Comment by jclouds-commentator** [comment]
to review your changes.

**Comment by nacx** [comment]
The previous comment is incomplete because the site CI build failed. The error is:

    HTTP/1.1 403 Forbidden; content: [{"userDisabled":{"code":403,"message":"User 'apachejclouds' is disabled."}}

@zack-shoylev could yoi have a look?

**Comment by jclouds-commentator** [comment]
to review your changes.

**Comment by jclouds-commentator** [comment]
to review your changes.

File: _posts/2016-08-15-arbitrary-cpu-ram.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+The previous Compute Service abstraction assumed that all providers had hardware profiles, but some providers such as [ProfitBricks](https://www.profitbricks.com/) or [ElasticHosts](https://www.elastichosts.com/) do not have the hardware profiles concept and the previous implementation provides a fixed configuration with a fixed (hardcoded) list just to conform the interface. The new implementation allows to use custom hardwares or both (when supported) hardware profiles and custom hardwares.
+
+### Providers supported
+There are several providers that support arbitrary values of CPU and RAM like Docker, ElasticHosts, Google Compute Engine, etc. The first available providers supported by the new feature are:
+
+* [Google Compute Engine](https://cloud.google.com/compute/)
+* [ProfitBricks](https://www.profitbricks.com/)
+
+To configure the new feature in other providers add a bind() to the ArbitraryCpuRamTemplateBuilderImpl class at the provider's context module:
```
Comment (nacx) [inline _posts/2016-08-15-arbitrary-cpu-ram.md]: Change the `ArbitraryCpuRamTemplateBuilderImpl `, to highlight it as code.

File: _posts/2016-08-15-arbitrary-cpu-ram.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+{% highlight Java %}
+bind(TemplateBuilderImpl.class).to(ArbitraryCpuRamTemplateBuilderImpl.class);
+{% endhighlight %}
+
+Also is necessary to modify the function that transform a node from the provider model to the portable model of jclouds, to include the new automatic hardwareId (if apply).
+
+### How to create custom hardwares
```
Comment (nacx) [inline _posts/2016-08-15-arbitrary-cpu-ram.md]: Don't say "create custom hardwares", as we're not creating anything. Better refer to the feature as "using custom hardwares".

File: _posts/2016-08-15-arbitrary-cpu-ram.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+compute.createNodesInGroup("jclouds", 1, template);
+{% endhighlight %}
+
+As some providers support also hardware profiles, and using them provides hardware optimization or in some cases cheaper pricings, when a provider support both, Template Builder first will look if some hardware profile matches the specified values.
+
+### Further development
+
+* **Support other providers**: add support for other providers such as [ElasticHosts](https://www.elastichosts.com/) and [Docker](https://www.docker.com/).
+* **Improve AutomaticHardwareSpec**: improve the AutomaticHardwareSpec with specific parsers for every parameter in order to support more custom parameters and some fields, like bootDisk and durable (part of volumes), that are currently hardcoded to true.
+* **Usage examples of the new features**: create examples of the new features in the jclouds-examples repo.
```
Comment (nacx) [inline _posts/2016-08-15-arbitrary-cpu-ram.md]: Add ti this list the remaining things for ProfitBricks:

* Add a custom implementation of the TemplateBuilderImpl to fail early when users don't set the minDisk.
* That custom implementation should also accept hardware IDs with the "automatic" prefix.

**Comment by nacx** [comment]
Thanks @ivanlomba!

@demobox Would you have some time to have a look at this before we publish it?

**Comment by jclouds-commentator** [comment]
to review your changes.

File: _posts/2016-08-15-arbitrary-cpu-ram.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+### Special thanks
+
+Special thanks to [Ignasi Barrera](https://github.com/nacx) for all the help, [Andrew Phillips](https://github.com/demobox) for code reviews and the rest of jclouds comunity.
```
Comment (demobox) [inline _posts/2016-08-15-arbitrary-cpu-ram.md]: Thanks, Ivan, but I think Ignasi should get the credit here - he did **much** more work than I did ;-)

**Comment by demobox** [comment]
Thanks for putting this together, @ivanlomba! Some general questions/comments:

* Who is the intended target audience here: people who want to _use_ these providers or people who want to _implement_ them? Or both? At the moment, we are mixing both types of information, and if we want to do that I think we should focus on the "use arbitrary CPU and RAM" first, since that's likely to be much more common.
* When we demonstrate how to use it, is there some guidance we can give about which mechanism (hardware ID vs. template builder) is preferred/recommended? Users generally find it easier to be given _one_ option, rather than multiple, especially if there isn't clear information about when to choose which option.
* The example of some providers requiring disk info: is there any way for a user to know that? Or do they have to try without and wait for an error message?
* We introduce "hardware profiles" as terms quite early on without explaining them - are we assuming they are sufficiently common that everyone (or, at least, all the likely readers) will know what they are?

/cc @nacx

**Comment by ivanlomba** [comment]
Thanks @demobox

>
Who is the intended target audience here: people who want to use these providers or people who want to implement them? Or both? At the moment, we are mixing both types of information, and if we want to do that I think we should focus on the "use arbitrary CPU and RAM" first, since that's likely to be much more common.

The main intention of the post is to give a general idea of the new feature and future work, specially how to use it, but also introducing how to extend it to other providers and commenting next steps. To avoid mixing information, I will put first the user guide and below the developers guide.

> When we demonstrate how to use it, is there some guidance we can give about which mechanism (hardware ID vs. template builder) is preferred/recommended? Users generally find it easier to be given one option, rather than multiple, especially if there isn't clear information about when to choose which option.

I think maybe explain a little bit the two options(clarifying the implications of using the hardwareId or "mins" option) can be better that only give a recommended one:

1. AutomaticId: the template builder checks if a hardware profile with exactly that
 hardwareId exist, if don't use the automatic.

2. Mins: the template builder checks if a hardware profile match with the provided minRam and minCores, if not use the automatic.

> The example of some providers requiring disk info: is there any way for a user to know that? Or do they have to try without and wait for an error message?

They have to try without know that.

> We introduce "hardware profiles" as terms quite early on without explaining them - are we assuming they are sufficiently common that everyone (or, at least, all the likely readers) will know what they are?

I think that a potential user of jclouds will understand it, but I can introduce the hardware profiles if you think that is necessary.

**Comment by jclouds-commentator** [comment]
to review your changes.

**Comment by demobox** [comment]
Thanks for all the changes, @ivanlomba! There are a bunch of small language things that I'm happy to do as a final pass, but from a content perspective I think we're pretty much there.

The one question I still have is whether we can give any kind of guidance about whether to use the builder or to construct an appropriate `automatic:...` ID string. We still mention both and give examples of both without recommending one or the other (whether in general or in particular cases).

If we feel they are really totally equivalent, perhaps point out that using the builder makes you slightly less prone to problems with string manipulation that you get when you construct the ID string yourself? Also, are there any helpers/utility objects (even if it's just something like `UriBuilder..addParam(...).addParam(...)...getQueryString()`) that can make constructing that string easier for users?

/cc @nacx

**Comment by nacx** [comment]
Thanks for reviewing @demobox!

@ivanlomba could you add some examples of using the existing utility class to build the automatic id string, as suggested?

If you add the change **today** and @demobox agrees, I'll merge the PR and publish the post. This would allow you to have a URL to add to the GSoC deliverables. @demobox are you OK with merging this early (given that the content is OK) for this purpose, even if you amend the syntax once publised?

**Comment by ivanlomba** [comment]
@nacex I added a new example using the AutomaticHardwareIdSpecBuilder.

@demobox I'm not sure about which option can be the best recommendation: one always creates a custom hardware (if user specify hardwareId) and the other first search for hardware profiles with similar values (minRam, minCores). For that reason I told you in my previous comment that I think that a better option was clarify the implications of any method.

**Comment by jclouds-commentator** [comment]
to review your changes.

**Comment by nacx** [comment]
It really depends on the use case. Users using the *minCores/Ram* methods don't really care about which concrete hardware profile is used as long as it satisfies those minimums. The current approach respects that: it tries to find an existing profile that meets the criteria and, if none is found, returns an *automatic* one with the given values.

Users using the *hardwareId* method want to use concrete profiles. Since automatic profiles don't exist but they need an ID in order to honor the Hardware and NodeMetadata interfaces, we use a parseable ID that can be used in the `TemplateBuilder.hardwareId` method too. This allows users to create deterministic code for automatic/explicit hardwares the same way they did before.

In the end it is about the use case, which shouldn't have changed with the introduction of these feature. The reasoning behind the choice of one or the other should remain the same.

Finally, regarding the first option (*minCores/Ram*), now providers will probably always find an existing hardware (and not fallback to the automatic behaviour). This is because in providers that don't really have hardware profiles we were returning mock values in the `ComputeService.listHardwares` method. Those should be changed to return an empty list, thus better leveraging the automatic hardware feature, but we've keeped them in this first version for backward compatibility (ther might be users coding against those mock hardwares).

**Comment by demobox** [comment]
> are you OK with merging this early (given that the content is OK) for this purpose, even if you amend the syntax once publised?

@nacx Absolutely! Sorry for the late response - only just seeing this :-(

> Users using the hardwareId method want to use concrete profiles. Since automatic profiles don't exist but they need an ID in order to honor the Hardware and NodeMetadata interfaces, we use a parseable ID that can be used in the TemplateBuilder.hardwareId method too.

Ah, thanks for clarifying. In that case, if I'm understanding the correctly the _only_ use case for a user _explicitly_ setting an `automatic:` ID would be if there's a matching hardware profile but the user does **not** want to use it, for some reason?

If there is **no** matching hardware profile, both methods will return the same for a provider that supports automatic profiles - is that correct? If so, I'd be tempted to recommend the `.minCores` method generally and point out (if not already done) that this _could_ cause the implementation to return a matching hardware profile if one happens to match those specs. For users that explicitly do **not** want an existing profile, we can recommend that they set the ID explicitly.

This, I think, comes close to respecting the origin of the ID as essentially being something we need to generate for consistency with the rest of the system. Would be curious to hear your thoughts on that, of course...but _after_ this is published, most likely ;-)

/cc @nacx @ivanlomba

**Comment by nacx** [comment]
Squashed, pushed to master as [386c1763](http://git-wip-us.apache.org/repos/asf/jclouds-site/commit/386c1763) and published!

@ivanlomba feel free to include the link in your deliverables as long as your commits to this repo and please, open a PR to include the last comments from @demobox.

I'm closing this PR now, but feel free to keep the discussion here if appropriate.

Thanks both!

**Comment by nacx** [comment]
I've just changed the date of the post. Please make sure you update the URL! It is the one for August 22.

@demobox no need to hurry for this at all; once amended we'll tweet the post from our Twitter account.

**Pull request closed without merging**